### PR TITLE
[MPS] Fix embedding_bag index bounds check off-by-one and clamp backward writes

### LINKED ISSUE #189971: [MPS] embedding_bag accepts index == num_embeddings and reads one row past the weight table
### 🐛 Describe the bug

The MPS `embedding_bag` weight-index bounds check is off by one: the kernel
uses `> num_weights` instead of `>= num_weights`, so index exactly equal to
`num_embeddings` silently reads one row past the weight table. Backward can
apply the same invalid row as an out-of-range gradient write.

```python
import torch

# Weight table = first 5 rows of a 6-row buffer, so the out-of-bounds row
# (index 5 == num_weights) lands on known adjacent data big[5] = [15,16,17].
big = torch.arange(18.0).reshape(6, 3)
W = big.to("mps")[:5]                     # num_weights = 5, valid idx 0..4
idx = torch.tensor([5], device="mps")     # == num_weights
offs = torch.tensor([0], device="mps")
out = torch.nn.functional.embedding_bag(idx, W, offs, mode="sum")
torch.mps.synchronize()
print("MPS returned:", out.cpu().tolist())  # [[15.0, 16.0, 17.0]] -- the row past the table
```

Index 6 correctly raises (`Index 0 is out of bounds: 6, range 0 to 5`),
pinning the failure to exactly the equality boundary. CPU raises for index 5.
Output on a source build of current main (2.14.0a0+git2662808); also
reproduces on 2.13.0 stable.

The guard is at
https://github.com/pytorch/pytorch/blob/26628087023e5102381849e9a27c3f09b6d485a4/aten/src/ATen/native/mps/kernels/EmbeddingBag.metal#L185-L194
-- `static_cast<uint32_t>(weight_idx) > num_weights` passes `weight_idx ==
num_weights`, and the subsequent row read uses it unchecked.

History: #163630 reported missing index validation and #168930 added this
check, but with `>` rather than `>=`; its regression test used index 6
against a 4-row table, which does not cover the equality boundary.

### Versions

PyTorch 2.14.0a0+git2662808 (source build of main) and 2.13.0; macOS 26.5.2 (arm64), Apple M5 Max, Python 3.12.13.

cc @kulinseth @malfet @DenisVieriu97 @jhavukainen @aditvenk @Isalia20

### PR DESCRIPTION
**Pull request by bogatyy**
The `embedding_bag` weight-index bounds check used `> num_weights` instead of `>=`, so index exactly equal to `num_embeddings` silently read one row past the weight table. #168930 added the check (for #163630) with the wrong comparator, and its regression test used index 6 against a 4-row table, missing the equality boundary. The error message also printed `num_weights` as the top of the valid range; it now prints `num_weights - 1`, and the existing `test_embedding_bag_out_of_bounds` regex is updated accordingly.

The backward kernels applied `weight_idx`/`max_indices` to gradient writes (and a weight read in the per-sample-weights kernel) unchecked, so the same invalid index also produced an out-of-bounds atomic gradient write before the forward's async error could surface. `num_weights` is added to the backward params and the kernels clamp, mirroring the check-then-clamp pattern used elsewhere in MPS indexing; the forward check remains the reporting point.

The new `test_embedding_bag_index_boundary` pins the boundary for sum/mean/max modes (index == num_embeddings raises `AcceleratorError`, num_embeddings - 1 matches CPU). All 55 embedding tests pass.

Related: #187572 adds offsets validation to the same kernel file; the diffs don't overlap.

Fixes #189971

*This PR was authored with assistance from Claude.*

### CODE REVIEW

**Comment by bogatyy** [comment]
@malfet gentle ping.

I kept the regression test in test_mps because it asserts the MPS out-of-bounds error contract specifically (an AcceleratorError raised at synchronize, since MPS reports device-side index errors asynchronously, unlike CPU's synchronous raise or CUDA's context-poisoning assert; the generic embedding_bag out-of-bounds test is already skipMPS'd for this reason). I dropped the part that re-checked a valid index against CPU, since that value contract is already covered by the device-generic consistency tests; what remains is the async rejection of index == num_embeddings and a guard that the last valid index is not over-rejected.
